### PR TITLE
Feat: g7-85

### PR DESCRIPTION
**Pull request by karacurt**
Issue: resolves https://linear.app/game7/issue/G7-85/verify-current-token-liquidity-and-pausability-of-the-contract

# 🐵 Description

This pull request introduces significant updates to the `GUnits` contract and its associated tests to support token migration with differing decimal values. Key changes include the addition of a `decimals` property, logic to handle token decimal adjustments during migrations, and comprehensive tests for these scenarios.

### Updates to `GUnits` Contract

* **Support for Token Decimals**:
  - Added a `decimals` state variable to store the decimal value of the current token. (`[contracts/upgradeables/games/GUnits.solR147](diffhunk://#diff-23ef7fab2ff47ca3e5a0688c510a942be09ef643c3c6516fb490d1b49c5671faR147)`)
  - Updated the initialization logic to set the `decimals` value based on the token's metadata. (`[contracts/upgradeables/games/GUnits.solR185](diffhunk://#diff-23ef7fab2ff47ca3e5a0688c510a942be09ef643c3c6516fb490d1b49c5671faR185)`)

* **Token Migration Logic**:
  - Modified the `setToken` function to include logic for handling token migrations with differing decimals. If the decimals differ, the `_rebalanceGUnitDecimals` internal function ensures the correct amount of the new token is deposited to back all existing G-Units. (`[contracts/upgradeables/games/GUnits.solL416-R448](diffhunk://#diff-23ef7fab2ff47ca3e5a0688c510a942be09ef643c3c6516fb490d1b49c5671faL416-R448)`)

* **Reserved Storage Adjustment**:
  - Reduced the reserved storage space (`__gap`) to accommodate the new `decimals` state variable. (`[contracts/upgradeables/games/GUnits.solL689-R716](diffhunk://#diff-23ef7fab2ff47ca3e5a0688c510a942be09ef643c3c6516fb490d1b49c5671faL689-R716)`)

### Updates to Tests

* **Expanded Token Migration Tests**:
  - Added tests to validate token migration scenarios, including migrations between tokens with differing decimal values (e.g., 6 to 18 decimals, 18 to 6 decimals, and 6 to 4 decimals). (`[test/gunits.test.tsL1400-R1536](diffhunk://#diff-5031254e21e229fcbd145431b2d460fc34268fd249cf43ce90e96ff664fd03bbL1400-R1536)`)
  - Verified correct deposit requirements and behavior during migrations, including edge cases where insufficient tokens are approved. (`[test/gunits.test.tsL1400-R1536](diffhunk://#diff-5031254e21e229fcbd145431b2d460fc34268fd249cf43ce90e96ff664fd03bbL1400-R1536)`)

* **General Improvements**:
  - Updated test utilities to support tokens with varying decimals by extending the `depositGUnits` function. (`[test/gunits.test.tsL72-R72](diffhunk://#diff-5031254e21e229fcbd145431b2d460fc34268fd249cf43ce90e96ff664fd03bbL72-R72)`)
  - Enhanced setup logic in tests to ensure proper initialization and pausing states for migration scenarios. (`[test/gunits.test.tsR1386-L1391](diffhunk://#diff-5031254e21e229fcbd145431b2d460fc34268fd249cf43ce90e96ff664fd03bbR1386-L1391)`)

These changes improve the flexibility and robustness of the `GUnits` contract, enabling seamless token migrations while maintaining the integrity of G-Unit values.

### CODE REVIEW

File: contracts/upgradeables/games/GUnits.sol [hunk 1]
```diff
@@ -372,12 +377,39 @@ contract GUnits is
...
-        if (_token == address(0)) {
+    // @param _previousTokenRecipient The address to send the previous token to
+    function setToken(address _newToken) external whenPaused onlyRole(DEV_CONFIG_ROLE) {
+        if (_newToken == address(0)) {
             revert AddressIsZero();
         }
-        token = _token;
-        emit TokenSet(_token);
+
+        if (totalSupply > 0) {
```
Comment (ogarciarevett) [inline contracts/upgradeables/games/GUnits.sol:386]: just readability if you check the totalSupply && the previousTokenBalance here, you don't need the second if.

**Review by ogarciarevett — COMMENTED** [review]
Small comment left

**Review by ogarciarevett — APPROVED** [review]
LGTM
